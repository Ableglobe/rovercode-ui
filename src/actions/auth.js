// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = isValidAuth => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});
