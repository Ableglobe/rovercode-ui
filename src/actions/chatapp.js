import Axios from "axios";

export const TOGGLE_FORMS = 'TOGGLE_FORMS';
export const SET_SESSION_ID = 'SET_SESSION_ID';
export const SET_CLIENT_ID = 'SET_CLIENT_ID';
export const SET_IS_SUPPORT_PROVIDER = 'SET_IS_SUPPORT_PROVIDER';


export const toggleForms = () => ({
  type: TOGGLE_FORMS,
});

export const setClientID = id => ({
  type: SET_CLIENT_ID,
  payload: id,
});

export const setSessionID = id => ({
  type: SET_SESSION_ID,
  payload: id,
});

export const setIsSupportProvider = () => ({
  type: SET_IS_SUPPORT_PROVIDER,
});

