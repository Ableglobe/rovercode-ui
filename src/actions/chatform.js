export const CATEGORY_SELECT_CHANGE = 'CATEGORY_SELECT_CHANGE';
export const BODY_INPUT_CHANGE = 'BODY_INPUT_CHANGE';
export const EXPERIENCE_SELECT_CHANGE = 'EXPERIENCE_SELECT_CHANGE'
export const SUBJECT_INPUT_CHANGE = 'SUBJECT_INPUT_CHANGE';

// CATEGORY STATES
export const FIRST_TIME = 0;
export const PRETTY_NEW = 1;
export const FOR_A_WHILE = 2;
export const EXPERT = 3;

//EXPERIENCE STATES
export const DEBUG = "DEBUG";
export const WHAT_NEXT = "WHAT_NEXT";
export const FUN_IDEAS = "FUN_IDEAS";


export const categorySelectChange = (value) =>({
    type: CATEGORY_SELECT_CHANGE,
    payload: value
})

export const subjectInputChange = (value) =>({
    type: SUBJECT_INPUT_CHANGE,
    payload: value
})

export const bodyInputChange = (value) =>({
    type: BODY_INPUT_CHANGE,
    payload: value
})

export const experienceSelectChange = (value) =>({
    type: EXPERIENCE_SELECT_CHANGE,
    payload: value
})

