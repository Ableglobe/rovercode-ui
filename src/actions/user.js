// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_USER = 'UPDATE_USER';

// action creators
export const updateUser = data => ({
  type: UPDATE_USER,
  payload: data,
});
