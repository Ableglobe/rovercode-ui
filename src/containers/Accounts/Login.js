import React, { Component, Fragment } from 'react';
import { Redirect } from 'react-router';
import {
  Button,
  Form,
  Grid,
  Icon,
  Message,
  Segment,
} from 'semantic-ui-react';
import axios from 'axios';
import URL from 'url-parse';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import { withCookies, Cookies } from 'react-cookie';

class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      socialError: false,
      basicError: false,
      basicSuccess: false,
      username: '',
      password: '',
    };
  }

  basicLogin = () => {
    const { cookies } = this.props;
    const { username, password } = this.state;

    return axios.post('/api/api-token-auth/', {
      username,
      password,
    })
      .then((response) => {
        cookies.set('auth_jwt', response.data.token, { path: '/' });
        this.setState({
          basicSuccess: true,
        });
      })
      .catch(() => {
        this.setState({
          basicError: true,
        });
      });
  }

  handlePasswordChange = (event) => {
    this.setState({
      password: event.target.value,
    });
  }

  handleUsernameChange = (event) => {
    this.setState({
      username: event.target.value,
    });
  }

  redirectToSocial = (element) => {
    const service = element.target.id;
    return axios.post(`/jwt/auth/social/${service}/auth-server/`)
      .then((response) => {
        const url = URL(response.data.url);
        const parsed = queryString.parse(url.query);
        const redirectUrl = URL(parsed.redirect_uri);
        redirectUrl.set('pathname', `/login/${service}/callback/`);
        parsed.redirect_uri = redirectUrl.toString();
        url.query = queryString.stringify(parsed);
        window.location.assign(url.toString());
      })
      .catch(() => {
        this.setState({
          socialError: true,
        });
      });
  }

  render() {
    const { basicError, basicSuccess, socialError } = this.state;

    return (
      <Fragment>
        {
          basicSuccess ? (
            <Redirect to="/" />
          ) : (null)
        }
        <Grid centered columns={16}>
          <Grid.Row>
            <p>
              You can sign up / sign in with one of your existing third-party accounts.
            </p>
          </Grid.Row>
          {
            socialError ? (
              <Grid.Row>
                <Message negative>
                  <Message.Header>
                    There was an error initiating social login.
                  </Message.Header>
                  <p>
                    Contact
                    {' '}
                    <a href="mailto:support@rovercode.com">
                      support@rovercode.com
                    </a>
                    {' '}
                    for help with this issue.
                  </p>
                </Message>
              </Grid.Row>
            ) : (null)
          }
          <Grid.Row>
            <Grid.Column width={8}>
              <Button id="google" floated="right" color="red" onClick={this.redirectToSocial}>
                <Icon name="google" />
                Sign in with Google
              </Button>
            </Grid.Column>
            <Grid.Column width={8}>
              <Button id="github" floated="left" color="black" onClick={this.redirectToSocial}>
                <Icon name="github" />
                Sign in with GitHub
              </Button>
            </Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <p>
              Don&apos;t have any of those accounts?
              {' '}
              <a href="/accounts/signup">
                Create a rovercode account.
              </a>
            </p>
          </Grid.Row>
          <Grid.Row>
            <p>
              Already have a rovercode account? Sign in here:
            </p>
          </Grid.Row>
          {
            basicError ? (
              <Grid.Row>
                <Message negative>
                  <Message.Header>
                    Invalid username or password.
                  </Message.Header>
                </Message>
              </Grid.Row>
            ) : (null)
          }
          <Grid.Row>
            <Grid.Column width={4}>
              <Segment raised secondary>
                <Form floated="left" onSubmit={this.basicLogin}>
                  <Form.Input required icon="user" iconPosition="left" placeholder="Username" onChange={this.handleUsernameChange} />
                  <Form.Input required icon="lock" iconPosition="left" type="password" placeholder="Password" onChange={this.handlePasswordChange} />
                  <Form.Button primary type="submit">
Sign In
                  </Form.Button>
                </Form>
              </Segment>
            </Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <p>
              <a href="/accounts/reset">
                Forgot Password?
              </a>
            </p>
          </Grid.Row>
        </Grid>
      </Fragment>
    );
  }
}

Login.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default withCookies(Login);
