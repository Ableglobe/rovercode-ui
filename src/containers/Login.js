import React, { Component, Fragment } from 'react';
import {
  Button,
  Grid,
  Icon,
  Image,
  Message,
} from 'semantic-ui-react';
import axios from 'axios';
import URL from 'url-parse';
import queryString from 'query-string';

import logoImage from '../assets/images/rovercode_logo.png';

export default class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      error: false,
    };
  }

  redirectToSocial = (element) => {
    const service = element.target.id;
    return axios.post(`/jwt/auth/social/${service}/auth-server/`)
      .then((response) => {
        const url = URL(response.data.url);
        const parsed = queryString.parse(url.query);
        const redirectUrl = URL(parsed.redirect_uri);
        redirectUrl.set('pathname', `/login/${service}/callback/`);
        parsed.redirect_uri = redirectUrl.toString();
        url.query = queryString.stringify(parsed);
        window.location.assign(url.toString());
      })
      .catch(() => {
        this.setState({
          error: true,
        });
      });
  }

  render() {
    const { error } = this.state;

    return (
      <Fragment>
        <Grid centered columns="equal">
          <Grid.Row>
            <Image src={logoImage} />
          </Grid.Row>
          <Grid.Row>
            <p>
              You can sign up / sign in with one of your existing third-party accounts.
            </p>
          </Grid.Row>
          {
            error ? (
              <Grid.Row>
                <Message negative>
                  <Message.Header>
                    There was an error initiating social login.
                  </Message.Header>
                  <p>
                    Contact
                    {' '}
                    <a href="mailto:support@rovercode.com">
                      support@rovercode.com
                    </a>
                    {' '}
                    for help with this issue.
                  </p>
                </Message>
              </Grid.Row>
            ) : (null)
          }
          <Grid.Row>
            <Grid.Column>
              <Button id="google" floated="right" color="red" onClick={this.redirectToSocial}>
                <Icon name="google" />
                Sign in with Google
              </Button>
            </Grid.Column>
            <Grid.Column>
              <Button id="github" floated="left" color="black" onClick={this.redirectToSocial}>
                <Icon name="github" />
                Sign in with GitHub
              </Button>
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </Fragment>
    );
  }
}
