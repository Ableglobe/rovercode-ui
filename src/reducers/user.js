import {
  UPDATE_USER,
  EDIT_USER_USERNAME,
  EDIT_USER_USERNAME_FULFILLED,
  EDIT_USER_USERNAME_REJECTED,
  EDIT_USER_PASSWORD,
  EDIT_USER_PASSWORD_FULFILLED,
  EDIT_USER_PASSWORD_REJECTED,
} from '../actions/user';

export default function user(
  state = {
    user_id: null,
    username: null,
    email: null,
    exp: null,
    isSocial: false,
    isEditingUsername: false,
    isEditingPassword: false,
    editUsernameError: null,
    editPasswordError: null,
  },
  action,
) {
  switch (action.type) {
    case UPDATE_USER:
      return {
        ...state,
        user_id: action.payload.user_id,
        username: action.payload.username,
        email: action.payload.email,
        exp: action.payload.exp,
        isSocial: action.payload.isSocial,
      };
    case EDIT_USER_USERNAME:
      return {
        ...state,
        isEditingUsername: true,
      };
    case EDIT_USER_USERNAME_FULFILLED:
      return {
        ...state,
        isEditingUsername: false,
        username: action.payload.username,
        email: action.payload.email,
      };
    case EDIT_USER_USERNAME_REJECTED:
      return {
        ...state,
        isEditingUsername: false,
        editUsernameError: action.payload,
      };
    case EDIT_USER_PASSWORD:
      return {
        ...state,
        isEditingPassword: true,
      };
    case EDIT_USER_PASSWORD_FULFILLED:
      return {
        ...state,
        isEditingPassword: false,
      };
    case EDIT_USER_PASSWORD_REJECTED:
      return {
        ...state,
        isEditingPassword: false,
        editPasswordError: action.payload,
      };
    default:
      return state;
  }
}
