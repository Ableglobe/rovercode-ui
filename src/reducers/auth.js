import { UPDATE_VALID_AUTH } from '../actions/auth';

export default function auth(
  state = {
    isValidAuth: true,
  },
  action,
) {
  switch (action.type) {
    case UPDATE_VALID_AUTH:
      return {
        ...state,
        isValidAuth: action.payload,
      };
    default:
      return state;
  }
}
