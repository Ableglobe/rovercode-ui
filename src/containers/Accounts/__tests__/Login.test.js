import React from 'react';
import { Redirect } from 'react-router';
import { Form, Message, MessageHeader } from 'semantic-ui-react';
import { shallow } from 'enzyme';
import { Cookies } from 'react-cookie';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import configureStore from 'redux-mock-store';
import jwtDecode from 'jwt-decode';

import { updateValidAuth } from '@/actions/auth';
import { updateUser } from '@/actions/user';

import Login from '../Login';

const mockStore = configureStore();
const store = mockStore();
store.dispatch = jest.fn();

const mock = new MockAdapter(axios);

const cookiesValues = { };
const cookies = new Cookies(cookiesValues);
const location = {
  pathname: '/login',
};

test('Login renders on the page with no errors', () => {
  const cookiesWrapper = shallow(<Login location={location} store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive();

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(Message).exists()).toBe(false);
});

test('Login redirects to social api on button click', async () => {
  const url = 'https://accounts.google.com/o/oauth2/auth?redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fauth%2Fsocial%2Fgoogle%2Fcallback';
  mock.reset();
  mock.onPost('/jwt/auth/social/google/auth-server/').reply(200, { url });
  window.location.assign = jest.fn();

  const element = {
    target: {
      id: 'google',
    },
  };

  const cookiesWrapper = shallow(<Login location={location} store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive();

  await wrapper.instance().redirectToSocial(element);

  expect(window.location.assign).toBeCalledWith(
    'https://accounts.google.com/o/oauth2/auth?redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Faccounts%2Flogin%2Fcallback%2Fgoogle',
  );
});

test('Login shows error message on api error', async () => {
  mock.reset();
  mock.onPost('/jwt/auth/social/google/auth-server/').timeout();
  window.location.assign = jest.fn();

  const element = {
    target: {
      id: 'google',
    },
  };

  const cookiesWrapper = shallow(<Login location={location} store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive();

  await wrapper.instance().redirectToSocial(element);
  wrapper.update();

  expect(wrapper.find(Message).exists()).toBe(true);
  expect(wrapper.find(MessageHeader).prop('children')).toBe('There was an error initiating social login.');
  expect(window.location.assign).not.toBeCalled();
});

test('Login redirects to root after basic login success', async () => {
  const token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxLCJ1c2VybmFtZSI6ImFkbWluIiwiZXhwIjoxNTQwMzQzMjIxLCJlbWFpbCI6ImFkbWluQGV4YW1wbGUuY29tIiwib3JpZ19pYXQiOjE1NDAzMzk2MjF9.tumcSSAbKeWXc2QDd7KFR9IGh3PCsyHnCe6JLSszWpc';
  const username = 'admin';
  const password = 'password';

  mock.reset();
  mock.onPost('/api/api-token-auth/', {
    username,
    password,
  }).reply(200, {
    token,
  });

  const cookiesWrapper = shallow(<Login location={location} store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive();

  wrapper.find(Form.Input).first().simulate('change', {
    target: {
      value: username,
    },
  });
  wrapper.find(Form.Input).last().simulate('change', {
    target: {
      value: password,
    },
  });

  await wrapper.instance().basicLogin();
  wrapper.update();

  expect(cookies.get('auth_jwt')).toBe(token);
  expect(wrapper.find(Redirect).exists()).toBe(true);
  expect(store.dispatch).toHaveBeenCalledWith(updateUser(jwtDecode(token)));
  expect(store.dispatch).toHaveBeenCalledWith(updateValidAuth(true));

  cookies.remove('auth_jwt');
});

test('Login shows error message after basic login failure', async () => {
  const username = 'admin';
  const password = 'password';

  mock.reset();
  mock.onPost('/api/api-token-auth/', {
    username,
    password,
  }).reply(400);

  const cookiesWrapper = shallow(<Login location={location} store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive();

  wrapper.find(Form.Input).first().simulate('change', {
    target: {
      value: username,
    },
  });
  wrapper.find(Form.Input).last().simulate('change', {
    target: {
      value: password,
    },
  });

  await wrapper.instance().basicLogin();
  wrapper.update();

  expect(wrapper.find(Message).exists()).toBe(true);
  expect(wrapper.find(MessageHeader).prop('children')).toBe('Invalid username or password.');
  expect(cookies.get('auth_jwt')).toBeUndefined();
});
