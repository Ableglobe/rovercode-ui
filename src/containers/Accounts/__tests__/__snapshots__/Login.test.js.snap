// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders on the page with no errors 1`] = `
<Fragment>
  <Grid
    centered={true}
    columns={16}
  >
    <GridRow>
      <p>
        <FormattedMessage
          defaultMessage="You can sign up / sign in with one of your existing third-party accounts."
          id="app.login.social"
          values={Object {}}
        />
      </p>
    </GridRow>
    <GridRow>
      <GridColumn
        width={8}
      >
        <Button
          as="button"
          color="red"
          floated="right"
          id="google"
          onClick={[Function]}
        >
          <Icon
            as="i"
            name="google"
          />
          <FormattedMessage
            defaultMessage="Sign in with Google"
            id="app.login.button_google"
            values={Object {}}
          />
        </Button>
      </GridColumn>
      <GridColumn
        width={8}
      >
        <Button
          as="button"
          color="black"
          floated="left"
          id="github"
          onClick={[Function]}
        >
          <Icon
            as="i"
            name="github"
          />
          <FormattedMessage
            defaultMessage="Sign in with GitHub"
            id="app.login.button_github"
            values={Object {}}
          />
        </Button>
      </GridColumn>
    </GridRow>
    <GridRow>
      <p>
        <FormattedMessage
          defaultMessage="Don't have any of those accounts?"
          id="app.login.social_question"
          values={Object {}}
        />
         
        <a
          href="/accounts/signup"
        >
          <FormattedMessage
            defaultMessage="Create a rovercode account."
            id="app.login.create"
            values={Object {}}
          />
        </a>
      </p>
    </GridRow>
    <GridRow>
      <p>
        <FormattedMessage
          defaultMessage="Already have a rovercode account? Sign in here:"
          id="app.login.account_question"
          values={Object {}}
        />
      </p>
    </GridRow>
    <GridRow>
      <GridColumn
        width={4}
      >
        <Segment
          raised={true}
          secondary={true}
        >
          <Form
            as="form"
            floated="left"
            onSubmit={[Function]}
          >
            <FormInput
              as={[Function]}
              control={[Function]}
              icon="user"
              iconPosition="left"
              onChange={[Function]}
              placeholder="Username"
              required={true}
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              icon="lock"
              iconPosition="left"
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
            />
            <FormButton
              as={[Function]}
              control={[Function]}
              primary={true}
              type="submit"
            >
              <FormattedMessage
                defaultMessage="Sign In"
                id="app.login.sign_in"
                values={Object {}}
              />
            </FormButton>
          </Form>
        </Segment>
      </GridColumn>
    </GridRow>
    <GridRow>
      <p>
        <a
          href="/accounts/reset"
        >
          <FormattedMessage
            defaultMessage="Forgot Password?"
            id="app.login.forgot"
            values={Object {}}
          />
        </a>
      </p>
    </GridRow>
  </Grid>
</Fragment>
`;
